<a href="https://colab.research.google.com/github/johanasev/Introducci-n-a-la-IA/blob/main/content/LAB%2002.03%20-%20PANDAS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# LAB 02.03 - Pandas

In [7]:
!wget --no-cache -O init.py -q https://raw.githubusercontent.com/rramosp/ai4eng.v1/main/content/init.py
import init; init.init(force_download=False); init.get_weblink()


replicating local resources


In [155]:
from local.lib.rlxmoocapi import submit, session
session.LoginSequence(endpoint=init.endpoint, course_id=init.course_id, lab_id="L02.03", varname="student");

logging in as johana.sevillano@udea.edu.co... please wait
using course session ai4eng.v1.udea.r3::2024.1
success!! you are logged in


In [ ]:
import numpy as np
import pandas as pd

## Task 1: Extract data

we have a dataframe of items with a price and, **sometimes**, with an extra column (`margin`).

You will have to complete a function that will **filter** the dataframe selecting the rows:

- whose price is > 100
- **OR** whose margin is >10, if the column `margin` is present

your function must return **A LIST** with the item ids of the selected rows

For instance, with the following dataframe

             price  category
    itemid                  
    56556   108.15       2.0
    73065    83.54       2.0
    36619   114.42       0.0
    73414    82.94       0.0
    13410   115.13       2.0
    66153    91.50       1.0
    77380    85.82       0.0
    73249    95.95       0.0
    11654   100.22       2.0
    11972    77.16       1.0

your must return the following list:

     [56556, 36619, 13410, 11654]
     
But if you get the following dataframe

             price  category  margin
    itemid                          
    39059    98.11       0.0   11.04
    19526    98.11       1.0   11.25
    78176    94.34       1.0   10.51
    50948   102.37       1.0   10.77
    12111    98.07       1.0    8.50
    56191    98.53       1.0   11.65
    38887    91.49       2.0   11.24
    77915   117.30       0.0    8.64
    55010    96.13       0.0    8.95
    45925    98.59       1.0   10.45
    
you must return the following list

    [39059, 19526, 78176, 50948, 56191, 38887, 77915, 45925]

In [64]:
def create_df(missing=False, n=10):
    itemid   = np.random.randint(100000, size=n)+1000
    category = np.random.randint(3, size=n)
    price    = np.round(np.random.normal(loc=100, scale=10, size=n),2)
    margin   = np.round(np.random.normal(loc=10, scale=1, size=n),2)

    if missing:
        nmissing = np.random.randint(len(price)//2)+2
        price[np.random.permutation(len(price))[:nmissing]] = np.nan

    d = pd.DataFrame(np.r_[[price, category, margin]].T, index=itemid, columns=["price", "category", "margin"])
    d.index.name="itemid"
    if np.random.random()>.5:
        d = d[d.columns[:2]]

    return d

In [67]:
import numpy as np
import pandas as pd

d = create_df()
d

price  category  margin
itemid                          
35447   104.41       2.0    9.74
94493   103.30       0.0    8.77
58913   107.87       0.0   10.30
86203   103.92       1.0    9.24
17071    93.25       2.0   10.71
80036   106.40       0.0   10.99
93306    98.40       1.0    9.83
40010    77.71       2.0   12.47
35100   105.47       2.0    8.87
49806   103.47       2.0    9.97

In [68]:
def select_items(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()

    filtro_df = df[(df['price']>100) | (df.get('margin', 0)>10)]

    result = filtro_df.index.tolist()

    return result

**manually check your answer**

In [82]:
d = create_df()
d

price  category  margin
itemid                          
68669   132.51       0.0   10.33
73117    77.02       0.0   10.60
75904   105.60       1.0   11.22
76881    85.84       1.0    9.34
66306   111.57       1.0   13.32
35199    86.90       1.0   10.05
2809     96.59       0.0   10.15
6141     90.90       1.0    9.59
75606   101.92       1.0   10.33
54049    90.59       2.0    9.92

In [83]:
select_items(d)

[68669, 73117, 75904, 66306, 35199, 2809, 75606]

**submit your code**

In [29]:
student.submit_task(globals(), task_id="task_01");

## Task 2: Group statistics

Complete the following function so that it returns a dataframe with the average, max and min **prices** per category.

For instance, for the following dataframe

             price  category  margin
    itemid                          
    17946    93.85       1.0   10.64
    61190    91.72       1.0    9.76
    39639   100.16       1.0   10.67
    17791   110.44       2.0    9.65
    7333    101.05       1.0    9.69
    77362   122.33       0.0   11.14
    92646   108.13       2.0   10.58
    27797    85.52       2.0   10.88
    31746    97.56       0.0    9.75
    12355   101.04       2.0    9.51
    
you should return the following dataframe

                 media  maximo  minimo
    categoria                         
    0         109.9450  122.33  97.56
    1          96.6950  101.05  91.72
    2         101.2825  110.44  85.52
    
observe that your result
- must not be a **multilevel** columnset.
- the column names and the index name must be **exactly** as in the example.
- the **index** must be of type **int**.

In [157]:
from os import stat
import numpy as np
import pandas as pd

def get_stats(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()

    stats_df = df.groupby('category')['price'].agg(['mean', 'max', 'min'])
    stats_df.columns = ['media', 'maximo', 'minimo']
    stats_df.index.name = 'categoria'
    stats_df.index = stats_df.index.astype(int)

    return stats_df

**manually check your answer**

In [150]:

d = create_df()
d

price  category  margin
itemid                          
65625    90.85       1.0   10.26
21070   101.58       0.0   11.47
76380    89.86       0.0   10.37
52444   100.72       0.0   10.37
92477   101.99       2.0    9.77
82304   107.88       0.0   12.03
28086    94.10       0.0   10.09
73689   101.01       1.0   10.34
63761    98.58       0.0    9.56
29578    93.84       2.0   10.59

In [158]:
get_stats(d)

media  maximo  minimo
categoria                           
0          98.786667  107.88   89.86
1          95.930000  101.01   90.85
2          97.915000  101.99   93.84

**submit your code**

In [159]:
student.submit_task(globals(), task_id="task_02");

## Task 3: Fill in missing data

Fill in the missing data in the **price** column with the following procedure:

- compute the mean and std of the avaialable prices
- sample from a normal distribution with the computed mean and std (see [`np.random.normal`])(https://numpy.org/doc/stable/reference/random/generated/numpy.random.normal.html), as many samples as missing data
- substitute the missing values with the samples

For instance, for this input dataframe:

             price  category  margin
    itemid                          
    18922      NaN       1.0   10.32
    69500   121.25       1.0   10.22
    76442    90.25       1.0   12.60
    33863   106.51       0.0   10.26
    15904    95.87       1.0   11.51
    41946   103.47       2.0    9.85
    85451    93.08       2.0    9.56
    70028   116.68       1.0    9.11
    26860      NaN       2.0    9.71
    12807    91.48       0.0    9.77
    
your solution might be similar to this (not exactly the same as you will be sampling data):


                 price  category  margin
    itemid                              
    18922    97.441188       1.0   10.32
    69500   121.250000       1.0   10.22
    76442    90.250000       1.0   12.60
    33863   106.510000       0.0   10.26
    15904    95.870000       1.0   11.51
    41946   103.470000       2.0    9.85
    85451    93.080000       2.0    9.56
    70028   116.680000       1.0    9.11
    26860   103.294843       2.0    9.71
    12807    91.480000       0.0    9.77
    
    
**WARN**: your function must not modify the original dataframe, make a copy of the input dataframe, fill the values in the copy and return it.

**HINT**: use the [`isna()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isna.html) method of a dataframe or a series.

In [ ]:
d = create_df(missing=True)
d

In [160]:
def fillna(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()

    mean_price = df['price'].mean()
    std_price = df['price'].std()
    missing_ind = df['price'].isna()
    num_missing = missing_ind.sum()

    if num_missing > 0:
      df_val = np.random.normal(mean_price, std_price, num_missing)
      df.loc[missing_ind, 'price'] = df_val

    return df

**check manually your code**

In [161]:
d = create_df(missing=True)
d

price  category  margin
itemid                          
65889   111.93       1.0   10.70
98975    92.71       0.0   10.60
29153      NaN       1.0   11.91
22852      NaN       0.0   11.01
71700      NaN       0.0    8.81
6307    103.50       0.0   11.13
33326   112.33       2.0    9.58
43296      NaN       2.0    9.04
48246      NaN       2.0    9.66
46077      NaN       1.0    8.73

In [162]:
fillna(d)

price  category  margin
itemid                              
65889   111.930000       1.0   10.70
98975    92.710000       0.0   10.60
29153   117.003781       1.0   11.91
22852    93.149577       0.0   11.01
71700   110.440911       0.0    8.81
6307    103.500000       0.0   11.13
33326   112.330000       2.0    9.58
43296   111.564026       2.0    9.04
48246   104.808757       2.0    9.66
46077   111.396134       1.0    8.73

**submit your code**

In [163]:
student.submit_task(globals(), task_id="task_03");